ouroboros-consensus-cardano/src/shelley/Ouroboros/Consensus/Shelley/Protocol/TPraos.hs

```diff
@@ -44,13 +60,17 @@ instance PraosCrypto c => ProtocolHeaderSupportsEnvelope (TPraos c) where
   pHeaderSize = fromIntegral . SL.bHeaderSize
   pHeaderBlockSize = SL.bsize . SL.bhbody
 
-  type EnvelopeCheckError _ = ChainPredicateFailure
-
-  envelopeChecks cfg lv hdr =
-    SL.chainChecks
-      maxPV
-      (SL.lvChainChecks lv)
-      (SL.makeHeaderView $ protocolHeaderView @(TPraos c) hdr)
+  type EnvelopeCheckError _ = TPraosEnvelopeError
+
+  envelopeChecks cfg checkpoints lv hdr = do
+    liftEither . first ChainPredicateFailure $
+      SL.chainChecks
+        maxPV
+        (SL.lvChainChecks lv)
+        (SL.makeHeaderView $ protocolHeaderView @(TPraos c) hdr)
+    whenJust (Map.lookup (pHeaderBlock hdr) checkpoints) $ \checkpoint ->
+      unless (checkpoint == pHeaderHash hdr) $
```

Comment:
same comment as other guard

Reply:
https://github.com/input-output-hk/ouroboros-consensus/pull/453#discussion_r1371394977

Reply:
Switched unless to when in 5c93fa0